nearcore/src/config.rs

```diff
@@ -1098,32 +1099,64 @@ pub fn get_config_url(chain_id: &String) -> String {
     )
 }
 
-pub fn download_file(url: &String, path: &Path, limit: usize) {
-    actix::System::new().block_on(async move {
-        let client = awc::Client::new();
-        let mut response = client.get(url).send().await.expect("Unable to download the file");
-        // IMPORTANT: limit specifies the maximum size of the genesis or config file
-        // In case where the genesis or config file is bigger than the specified
-        // limit Overflow Error is thrown
-        let body = response
-            .body()
-            .limit(limit)
-            .await
-            .expect("File is bigger than specified limit. Please make the limit higher.");
-
-        std::fs::write(&path, &body).expect("Failed to create / write a file.");
+#[derive(thiserror::Error, Debug)]
+pub enum FileDownloadError {
+    #[error("Failed to download the file: {0}")]
+    HttpError(#[from] hyper::Error),
+    #[error("Failed to open file: {0}")]
+    OpenError(std::io::Error),
+    #[error("Failed to write to file: {0}")]
+    WriteError(std::io::Error),
+    #[error("Failed to rename file: {0}")]
+    RenameError(std::io::Error),
+    #[error("Invalid URI: {0}")]
+    UriError(#[from] hyper::http::uri::InvalidUri),
+    #[error("Failed to remove the temporary file after failure: {0}, {1}")]
+    RemoveTemporaryFileError(std::io::Error, Box<FileDownloadError>),
+}
+
+pub fn download_file(url: &String, path: &Path) -> Result<(), FileDownloadError> {
+    return actix::System::new().block_on(async move {
+        let https_connector = hyper_tls::HttpsConnector::new();
+        let client = hyper::Client::builder().build::<_, hyper::Body>(https_connector);
+
+        let uri = url.parse()?;
+        let mut resp = client.get(uri).await?;
+
+        // To avoid partially downloaded files, we first download the file to a temporary *.swp file
+        // and rename it once the download is finished.
+        let tmp_path = path.with_extension("swp");
+        let mut tmp_file = File::create(&tmp_path).map_err(FileDownloadError::OpenError)?;
+
+        let process_tmp_file = async {
+            while let Some(next_chunk_result) = resp.data().await {
+                let next_chunk = next_chunk_result?;
+                tmp_file.write_all(next_chunk.as_ref()).map_err(FileDownloadError::WriteError)?;
```

Comment:
While we are at it, we might also make it async-ready since `std::fs` is actually blocking.

P.S. I am surprised that there is no good crate dealing with downloading files well; I have found the following crates:

* https://crates.io/crates/siwi-download - async `write_all` (:heavy_check_mark:), support for resuming downloads (:heavy_check_mark:), and progress-bars (:heavy_check_mark:), but code-wise I feel it is quite a hack project rather than a well-supported library
* https://crates.io/crates/download-async - it uses blocking `write_all` operation, which is a red flag to me
* https://crates.io/crates/downloader - it uses blocking `write_all` operation, which is a red flag to me

Reply:
@frol why not https://docs.rs/async-std/1.10.0/async_std/fs/struct.File.html? It's a drop in replacement for `std::fs::File`

Reply:
@frol 
> While we are at it, we might also make it async-ready since std::fs is actually blocking.

Makes sense, done! Used async-std as per @ChaoticTempest's suggestion.

@ChaoticTempest 
> why not https://docs.rs/async-std/1.10.0/async_std/fs/struct.File.html? It's a drop in replacement for std::fs::File

I think what @frol actually wanted is a decent e2e library that would **download a file**, not a low-level async library/http client. But since such a thing does not exist, I used `async-std` as per your suggestion.

Reply:
Async-std is another runtime, so we should use tokio::fs::File: https://docs.rs/tokio/1.14.0/tokio/fs/struct.File.html

Reply:
Done